metal/src/command_buffer.rs

```
@@ -0,0 +1,157 @@
+
+    sample_buffer_attachment_descriptor.set_sample_buffer(counter_sample_buffer);
+    sample_buffer_attachment_descriptor.set_start_of_encoder_sample_index(0);
+    sample_buffer_attachment_descriptor.set_end_of_encoder_sample_index(1);
+}
+
+#[derive(Debug, Clone)]
+pub struct MetalProfiler {
+    device: Device,
+    profile_buffers: HashMap<usize, Vec<Buffer>>,
+    current_node_id: Option<usize>,
+    counter_sample_buffer: CounterSampleBuffer,
+    compute_pass_descriptor: ComputePassDescriptor,
+}
+
+impl MetalProfiler {
+    pub fn new(device: Device) -> Self {
+        let compute_pass_descriptor = ComputePassDescriptor::new();
+
+        let counter_sample_buffer_desc = CounterSampleBufferDescriptor::new();
+        counter_sample_buffer_desc.set_storage_mode(metal::MTLStorageMode::Shared);
+        counter_sample_buffer_desc.set_sample_count(NUM_SAMPLES);
+        let counter_sets = device.counter_sets();
+        let timestamp_counter = counter_sets.iter().find(|cs| cs.name() == "timestamp");
+
+        counter_sample_buffer_desc
+            .set_counter_set(timestamp_counter.expect("No timestamp counter found"));
+
+        let counter_sample_buffer =
+            device.new_counter_sample_buffer_with_descriptor(&counter_sample_buffer_desc).unwrap();
+
+        handle_compute_pass_sample_buffer_attachment(
+            compute_pass_descriptor,
+            &counter_sample_buffer,
+        );
+
+        Self {
+            device: device.to_owned(),
+            profile_buffers: HashMap::new(),
+            current_node_id: None,
+            counter_sample_buffer,
+            compute_pass_descriptor: compute_pass_descriptor.to_owned(),
+        }
+    }
+
+    pub fn add_node_entry(&mut self, node_id: usize) {
+        self.profile_buffers.insert(node_id, vec![]);
+        self.current_node_id = Some(node_id);
+    }
+
+    pub fn add_buffer(&mut self, buffer: Buffer) {
+        let current_node_id = &self.current_node_id.unwrap();
```

unwrap should be avoid. We could return an Error instead with TractResult:

```
self.current_node_id.ok_or_else(|| anyhow!("Metal profile doesn't have any current node id which is unexpected while adding the sampling buffer"))?;
```